### PR TITLE
Configure RAG for GKE Autopilot

### PR DESCRIPTION
**Pull request by artemvmin**
Update RAG dependencies to use GKE Autopilot:
- HF TGI
- Frontend
- KubeRay
- Jupyter

Update README with Autopilot instructions.

### CODE REVIEW

File: applications/rag/README.md [hunk 1]
```diff
@@ -24,15 +24,25 @@ Preinstall the following on your computer:
...
+CLUSTER_NAME=rag-cluster
+CLUSTER_REGION=us-central1
+```
+
+1. This sample assumes a GKE cluster already exists. If not, we recommend creating an Autopilot cluster on the latest available version:
+```
+gcloud container clusters create-auto $CLUSTER_NAME --location $CLUSTER_REGION --cluster-version 1.29.1-gke.1589000
+```
+
+To create a GKE Standard cluster, please refer to our [LLM on GKE guide](https://cloud.google.com/kubernetes-engine/docs/tutorials/serve-multiple-gpu#create_cluster_node_pool) for cluster configuration parameters.
```
Comment (imreddy13) [inline applications/rag/README.md:39]: I'm assuming you tested it works E2E /pods come up with this create command for standard

File: applications/rag/README.md [hunk 2]
```diff
@@ -41,7 +51,12 @@ gcloud container node-pools create g2-standard-24 --cluster <cluster-name> \
  --node-locations $REGION-a,$REGION-b --location=$REGION
 ```
 
-4. Enable additional required APIs:
+2. Ensure your k8s client is using the correct cluster by running:
```
Comment (imreddy13) [inline applications/rag/README.md]: numbering is wrong

File: applications/rag/README.md [hunk 3]
```diff
@@ -41,6 +51,11 @@ gcloud container node-pools create g2-standard-24 --cluster <cluster-name> \
  --node-locations $REGION-a,$REGION-b --location=$REGION
 ```
 
+3. Ensure your k8s client is using the correct cluster by running:
+```
+gcloud container clusters get-credentials $CLUSTER_NAME --location $CLUSTER_REGION
+```
+
 4. Enable additional required APIs:
```
Comment (imreddy13) [inline applications/rag/README.md:59]: Can you delete this? we do it by default now: https://github.com/GoogleCloudPlatform/ai-on-gke/blob/main/modules/cloudsql/main.tf#L18

File: applications/rag/README.md [hunk 4]
```diff
@@ -41,6 +51,11 @@ gcloud container node-pools create g2-standard-24 --cluster <cluster-name> \
  --node-locations $REGION-a,$REGION-b --location=$REGION
 ```
 
+3. Ensure your k8s client is using the correct cluster by running:
```
Comment (imreddy13) [inline applications/rag/README.md:54]: i think this is only needed for the verify setup step, should we move it down?

File: applications/rag/README.md [hunk 5]
```diff
@@ -74,22 +89,27 @@ This filter can auto fetch the templates in your project. Please refer to the fo
...
 4. Verify the inference server is setup:
-* Set up port forward
+* Start port forwarding
 ```
-kubectl port-forward -n <namespace> deployment/mistral-7b-instruct 8080:8080 &
+kubectl port-forward -n $NAMESPACE deployment/mistral-7b-instruct 8080:8080
 ```
 
-* Try a few prompts:
+* In a new terminal, try a few prompts:
```
Comment (imreddy13) [inline applications/rag/README.md:112]: why?

File: applications/rag/README.md [hunk 6]
```diff
@@ -106,19 +126,19 @@ EOF
 * At the end of the smoke test with the TGI server, close the port forward for 8080.
```
Comment (imreddy13) [inline applications/rag/README.md:126]: nit: add steps for how to close port forward

File: modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+# sure to refer to the Helm chart version of interest to you!
+#
+# Introduction to YAML:     https://www.youtube.com/watch?v=cdLNKUoMc6c
+# Chart config reference:   https://zero-to-jupyterhub.readthedocs.io/en/stable/resources/reference.html
+# Chart default values:     https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/HEAD/jupyterhub/values.yaml
+# Available chart versions: https://jupyterhub.github.io/helm-chart/
+hub:
+  image:
+    name: us-docker.pkg.dev/ai-on-gke/jupyterhub-authentication-class/jupyter-auth-class
+    tag: latest
```
Comment (imreddy13) [inline modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:27]: [unrelated to this PR]: @blackzlq should we switch this to a different tag?

File: modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+  config:
+    JupyterHub:
+      authenticator_class: ${authenticator_class}
+    GCPIAPAuthenticator:
```
Comment (imreddy13) [inline modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:31]: @blackzlq @artemvmin We need to test AP with IAP

Comment (imreddy13) [reply modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:31]: See https://github.com/GoogleCloudPlatform/ai-on-gke/pull/250

File: modules/jupyter/main.tf [hunk 9]
```diff
@@ -113,24 +113,41 @@ resource "helm_release" "jupyterhub" {
   namespace        = var.namespace
   create_namespace = !contains(data.kubernetes_all_namespaces.allns.namespaces, var.namespace)
   cleanup_on_fail  = "true"
-  # timeout increased to support autopilot scaling resources, and give enough time to complete the deployment 
-  timeout          = 1200
+  # This timeout is sufficient and ensures terraform doesn't hang for 20 minutes on error.
```
Comment (imreddy13) [inline modules/jupyter/main.tf:116]: is the timeout sufficient for both AP and standard?

Comment (imreddy13) [reply modules/jupyter/main.tf:116]: @umeshkumhar @blackzlq wasnt the reason for this because cert provisioning takes 20 mins?

File: modules/jupyter/main.tf [hunk 10]
```diff
@@ -113,24 +113,41 @@ resource "helm_release" "jupyterhub" {
...
-  values = [
-    templatefile("${path.module}/jupyter_config/config-selfauth.yaml", {
+  values = var.autopilot_cluster ? [ templatefile("${path.module}/jupyter_config/config-selfauth-autopilot.yaml", {
       password            = var.add_auth ? "dummy" : random_password.generated_password[0].result
       project_id          = var.project_id
       project_number      = data.google_project.project.number
 
       # Support legacy image.
-      service_id          = ""#var.add_auth ? (data.google_compute_backend_service.jupyter-ingress[0].generated_id != null ? data.google_compute_backend_service.jupyter-ingress[0].generated_id : "no-id-yet") : "no-id-yet"
+      service_id          = "" # TODO(umeshkumhar): var.add_auth ? (data.google_compute_backend_service.jupyter-ingress[0].generated_id != null ? data.google_compute_backend_service.jupyter-ingress[0].generated_id : "no-id-yet") : "no-id-yet"
```
Comment (imreddy13) [inline modules/jupyter/main.tf:126]: describe what the TODO is?

File: modules/kuberay-cluster/kuberay-autopilot-values.yaml [hunk 11]
```diff
@@ -107,6 +107,7 @@ head:
     gke-gcsfuse/ephemeral-storage-limit: 20Gi
   nodeSelector:
     iam.gke.io/gke-metadata-server-enabled: "true"
+    cloud.google.com/compute-class: "Accelerator"
     cloud.google.com/gke-accelerator: "nvidia-tesla-t4"
```
Comment (imreddy13) [inline modules/kuberay-cluster/kuberay-autopilot-values.yaml:111]: needs to be l4?

File: modules/kuberay-cluster/kuberay-autopilot-values.yaml [hunk 12]
```diff
@@ -156,108 +157,15 @@ worker:
   # If you want to disable the default workergroup
   # uncomment the line below
   disabled: true
-  groupName: workergroup
```
Comment (imreddy13) [inline modules/kuberay-cluster/kuberay-autopilot-values.yaml:159]: why'd we wipe this?

File: modules/kuberay-cluster/kuberay-autopilot-values.yaml [hunk 13]
```diff
@@ -391,8 +207,10 @@ additionalWorkerGroups:
       gke-gcsfuse/memory-limit: 20Gi
       gke-gcsfuse/ephemeral-storage-limit: 20Gi
     nodeSelector:
+      cloud.google.com/compute-class: "Performance"
+      cloud.google.com/machine-family: "c3"
+      cloud.google.com/gke-ephemeral-storage-local-ssd: "true"
       iam.gke.io/gke-metadata-server-enabled: "true"
-      cloud.google.com/gke-accelerator: "nvidia-tesla-t4"
```
Comment (imreddy13) [inline modules/kuberay-cluster/kuberay-autopilot-values.yaml:395]: dont we still need to pick L4 somewhere for AP?

File: modules/kuberay-cluster/kuberay-autopilot-values.yaml [hunk 14]
```diff
@@ -438,12 +256,12 @@ additionalWorkerGroups:
         - mountPath: /fluent-bit/etc/
           name: fluentbit-config
 
-  largeGroup:
+  gpuGroup:
```
Comment (imreddy13) [inline modules/kuberay-cluster/kuberay-autopilot-values.yaml:259]: how does this work/ where is gpuGroup used?

File: modules/kuberay-cluster/kuberay-values.yaml [hunk 15]
```diff
@@ -236,85 +236,14 @@ worker:
         name: ray-logs
```
Comment (imreddy13) [inline modules/kuberay-cluster/kuberay-values.yaml:236]: same, wipe all references to T4 if we're updating cluster creation instructions to not include T4 nodepool. if you want, make this part a different PR so it's well tested. we need to ensure the ray workers come up on both standard and AP on L4s.

File: modules/kuberay-monitoring/config/frontend-deployment.yaml [hunk 16]
```diff
@@ -15,7 +15,7 @@
 apiVersion: apps/v1
 kind: Deployment
 metadata:
-  name: frontend
+  name: kuberay-frontend
```
Comment (imreddy13) [inline modules/kuberay-monitoring/config/frontend-deployment.yaml:18]: nit: kuberay-monitoring-frontend

File: modules/kuberay-monitoring/config/frontend-service.yaml [hunk 17]
```diff
@@ -15,7 +15,7 @@
 apiVersion: v1
 kind: Service
 metadata:
-  name: frontend
+  name: kuberay-frontend
```
Comment (imreddy13) [inline modules/kuberay-monitoring/config/frontend-service.yaml:18]: same

File: tutorials/hf-tgi/mistral-quantized-deploy.yaml
Comment (imreddy13) [inline tutorials/hf-tgi/mistral-quantized-deploy.yaml:1]: i think this file is unused - delete it?

File: applications/rag/main.tf
Comment (imreddy13) [inline applications/rag/main.tf:1]: update .tfvars to expose AP cluster as an option?

File: applications/rag/README.md (hunk 1 above)
Comment (artemvmin) [reply applications/rag/README.md]: Seems this was confusing. I only meant it as a guide for setting up a cluster with GPUs. Removed and split out the Autopilot section, as you recommended.

File: applications/rag/README.md (hunk 2 above)
Comment (artemvmin) [reply applications/rag/README.md]: Fixed.

File: applications/rag/README.md (hunk 3 above)
Comment (artemvmin) [reply applications/rag/README.md]: Done.

File: applications/rag/README.md (hunk 4 above)
Comment (artemvmin) [reply applications/rag/README.md]: Done.

File: applications/rag/README.md (hunk 5 above)
Comment (artemvmin) [reply applications/rag/README.md:116]: Background processes are not the most user friendly thing. Most people will know how to create to terminals. Even Google Cloud Console provides tabs for this purpose.

File: modules/kuberay-cluster/kuberay-values.yaml (hunk 15 above)
Comment (artemvmin) [reply modules/kuberay-cluster/kuberay-values.yaml:236]: Leaving this as-is and reverting README changes to Standard mode. I have not tested this on GKE Standard.

File: modules/kuberay-monitoring/config/frontend-deployment.yaml (hunk 16 above)
Comment (artemvmin) [reply modules/kuberay-monitoring/config/frontend-deployment.yaml]: Done.

File: modules/kuberay-monitoring/config/frontend-service.yaml (hunk 17 above)
Comment (artemvmin) [reply modules/kuberay-monitoring/config/frontend-service.yaml]: Also done.

File: tutorials/hf-tgi/mistral-quantized-deploy.yaml
Comment (artemvmin) [reply tutorials/hf-tgi/mistral-quantized-deploy.yaml:1]: I think so, too. Done.

File: applications/rag/main.tf
Comment (artemvmin) [reply applications/rag/main.tf:1]: This option is not specified by the user. It's pulled from the cluster data: https://github.com/GoogleCloudPlatform/ai-on-gke/blob/16894f8321220204ab02e6c451988cb60a7771e6/applications/rag/main.tf#L28

File: applications/rag/README.md (hunk 6 above)
Comment (artemvmin) [reply applications/rag/README.md]: Easy with two terminals. Done.

File: modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml (hunk 8 above)
Comment (artemvmin) [reply modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:31]: I assume this is also unrelated to this PR. We'll test once IAP is complete.

File: modules/jupyter/main.tf (hunk 9 above)
Comment (artemvmin) [reply modules/jupyter/main.tf:116]: This is where documentation is important. I was going off the original comment, which is incorrect, as Ray Autoscaler provisions on demand and not during deployment time.

Can we bump this later once IAP is fully implemented?

File: modules/jupyter/main.tf (hunk 10 above)
Comment (artemvmin) [reply modules/jupyter/main.tf:126]: I actually don't know what this comment is meant to accomplish. I'm just assigning it for Umesh to investigate. I also left a comment here: https://github.com/GoogleCloudPlatform/ai-on-gke/commit/088d7827902ff541474149ef4ae8ee340d1b74d1#r139178350

File: modules/kuberay-cluster/kuberay-autopilot-values.yaml (hunk 11 above)
Comment (artemvmin) [reply modules/kuberay-cluster/kuberay-autopilot-values.yaml]: On second thought, this is the head node and shouldn't be running an expensive GPU at all. The GPU worker group will be scaled up and down based on demand from scheduled Ray jobs.

File: modules/kuberay-cluster/kuberay-autopilot-values.yaml (hunk 12 above)
Comment (artemvmin) [reply modules/kuberay-cluster/kuberay-autopilot-values.yaml:159]: Because Ray Autoscaler and worker groups are disabled on the GKE Standard setup.

File: modules/kuberay-cluster/kuberay-autopilot-values.yaml (hunk 14 above)
Comment (artemvmin) [reply modules/kuberay-cluster/kuberay-autopilot-values.yaml:258]: Ray Autoscaler scales these groups automatically based on demand. My current understanding is that it is able to detect requirements for each Ray job and choose the correct workers for the job.

File: modules/kuberay-cluster/kuberay-autopilot-values.yaml (hunk 13 above)
Comment (artemvmin) [reply modules/kuberay-cluster/kuberay-autopilot-values.yaml:395]: Yes, in the next worker group. For RAG, our Ray job will likely scale the gpuGroup. However, other applications may prefer CPU.

File: modules/jupyter/main.tf (hunk 10 above)
Comment (umeshkumhar) [reply modules/jupyter/main.tf:126]: This will be handled by Liqin PR (https://github.com/GoogleCloudPlatform/ai-on-gke/pull/250).. wil be merged soon.

**Review by imreddy13 — APPROVED** [review]
